module/zfs/vdev_raidz_math_neon2.c

```diff
@@ -0,0 +1,109 @@
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (C) 2016 Romain Dolbeau. All rights reserved.
+ */
+
+#include <sys/isa_defs.h>
+
+#if defined(__aarch64__)
+
+#include "vdev_raidz_math_neon_common.h"
+
+#define	GEN_P_DEFINE()	GEN_X_DEFINE_ALL()
+#define	GEN_P_STRIDE		8
+#define	GEN_P_P			0, 1, 2, 3, 4, 5, 6, 7
+
+#define	GEN_PQ_DEFINE()	GEN_X_DEFINE_ALL()
+#define	GEN_PQ_STRIDE		4
+#define	GEN_PQ_D		0, 1, 2, 3
+#define	GEN_PQ_P		4, 5, 6, 7
+#define	GEN_PQ_Q		8, 9, 10, 11
+
+#define	GEN_PQR_DEFINE()	GEN_X_DEFINE_ALL()
+#define	GEN_PQR_STRIDE		4
+#define	GEN_PQR_D		0, 1, 2, 3
+#define	GEN_PQR_P		4, 5, 6, 7
+#define	GEN_PQR_Q		8, 9, 22, 23
+#define	GEN_PQR_R		24, 25, 26, 27
+
+#define	REC_P_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_P_STRIDE		4
+#define	REC_P_X			0, 1, 2, 3
+
+#define	REC_Q_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_Q_STRIDE		4
+#define	REC_Q_X			0, 1, 2, 3
+
+#define	REC_R_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_R_STRIDE		4
+#define	REC_R_X			0, 1, 2, 3
+
+#define	REC_PQ_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_PQ_STRIDE		4
+#define	REC_PQ_X		0, 1, 2, 3
+#define	REC_PQ_Y		4, 5, 6, 7
+#define	REC_PQ_D		8, 9, 22, 23
+
+#define	REC_PR_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_PR_STRIDE		4
+#define	REC_PR_X		0, 1, 2, 3
+#define	REC_PR_Y		4, 5, 6, 7
+#define	REC_PR_D		8, 9, 22, 23
+
+#define	REC_QR_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_QR_STRIDE		4
+#define	REC_QR_X		0, 1, 2, 3
+#define	REC_QR_Y		4, 5, 6, 7
+#define	REC_QR_D		8, 9, 22, 23
+
+#define	REC_PQR_DEFINE()	GEN_X_DEFINE_ALL()
+#define	REC_PQR_STRIDE		4
+#define	REC_PQR_X		0, 1, 2, 3
+#define	REC_PQR_Y		4, 5, 6, 7
+#define	REC_PQR_Z		8, 9, 22, 23
+#define	REC_PQR_D		24, 25, 26, 27
+#define	REC_PQR_XS		24, 25, 26, 27
+#define	REC_PQR_YS		28, 29, 30, 31
+
+
+#include <sys/vdev_raidz_impl.h>
+#include "vdev_raidz_math_impl.h"
+
+DEFINE_GEN_METHODS(neon2);
+DEFINE_REC_METHODS(neon2);
+
+static boolean_t
+raidz_will_neon2_work(void)
+{
+	return (B_TRUE); // __arch64__ requires NEON
+}
+
+const raidz_impl_ops_t vdev_raidz_neon2_impl = {
+	.init = NULL,
+	.fini = NULL,
+	.gen = RAIDZ_GEN_METHODS(neon2),
+	.rec = RAIDZ_REC_METHODS(neon2),
+	.is_supported = &raidz_will_neon2_work,
+	.name = "neon2"
```

Comment:
The actual performance in NEON will vary between CPU implementations, and those are more varied than in the x86-64 world. More unrolling should be at least as fast, but as a trade-off you get a potentially longer tail loop for blocks of a sub-optimal size. So in theory, one should pick the smallest unrolling that get maximum throughput... not easy.
I suspect neon2 would be a good default in most cases.
I don't have hard number to supply unfortunately, the systems where I have kernel headers to compile are under NDA :-(
I have access to a Jetson TX1 as well, but when I do 'make deb' (it's ubuntu-based), I get this ultimately:
#####
make[1]: Leaving directory `/home/ubuntu/spl'
name=spl; \
        version=0.7.0-rc1; \
        arch=`rpm -qp ${name}-${version}.src.rpm --qf %{arch} | tail -1`; \
        pkg1=${name}-${version}.${arch}.rpm; \
        fakeroot alien --bump=0 --scripts --to-deb $pkg1; \
        rm -f $pkg1
spl-0.7.0-rc1.aarch64.rpm is for architecture aarch64 ; the package cannot be built on this system
#####
Weird...